lib/kcerts/ssh.go

```diff
@@ -0,0 +1,125 @@
+package kcerts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string) error {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return err
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if _, err := os.Stat(sshDir); os.IsNotExist(err) {
+		fmt.Printf("ssh directory %s does not exist, attempting to create it", sshDir)
+		err = os.Mkdir(sshDir, 0700)
+		if err != nil {
+			return fmt.Errorf("could not create dir %s err: %w", sshDir, err)
+		}
+	}
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	if _, err := os.Stat(knownHosts); os.IsNotExist(err) {
+		fmt.Printf("ssh authorized hosts file %s does not exist, creating it now \n", knownHosts)
+		err = os.Mkdir(knownHosts, 0644)
+		if err != nil {
+			return fmt.Errorf("could not create file %s: %w", knownHosts, err)
+		}
+	}
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	existingKnownHostsContent, err := ioutil.ReadFile(knownHosts)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_APPEND|os.O_WRONLY, 0644)
```

Comment:
here is the critical part: you must be able to write the file, and create the directory if it does not exist. I would implement it like this (note os.O_CREAT, and 0600):

```
knownHostsFile, err := os.OpenFile(knownHosts, os.O_APPEND|os.O_WRONLY|os.O_CREAT, 0600)
if err != nil {
  if !os.IsNotExist(err) {
    return fmt.Errorf("could not create known_hosts file: %w", err)
  }
  if err := os.Mkdir(knownHosts, 0700); err != nil {
    return fmt.Errorf("could not create .ssh directory: %w", err)
  }
}
```

(open returns ENOENT, equivalent to NotExist, if some of the path components do not exist, https://play.golang.org/p/sesFoHhvFyz, and `man 2 open`)

Reply:
I've implemented this in a slightly different fashion that requires less nesting.